Remotion/Data/DomainObjects.UnitTests/IntegrationTests/TableValuedParametersIntegrationTest.cs

```diff
@@ -0,0 +1,318 @@
+  public void Contains_WithSetOfT_UsesTableTypeWithUniqueConstraint ()
+  {
+    // set up a set with duplicates to get a database exception when we insert these duplicates into a table with unique constraint
+    var possibleItems = SetupCollection<ISet<Guid>>([(Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value]);
+
+    var query = QueryFactory.CreateQuery(
+        new QueryDefinition("Test", TestDomainStorageProviderDefinition, "SELECT * FROM [Order] o WHERE o.[ID] IN (SELECT [Value] FROM @1)", QueryType.Collection),
+        new QueryParameterCollection { { "@1", possibleItems } });
+
+    Assert.That(() => TestableClientTransaction.QueryManager.GetCollection<Order>(query),
+        Throws.InstanceOf<RdbmsProviderException>()
+            .With.Message.Contains("Error while executing SQL command: Violation of UNIQUE KEY constraint")
+            .With.Message.Contains("Cannot insert duplicate key in object 'dbo.@1'. The duplicate key value is (83445473-844a-4d3f-a8c3-c27f8d98e8ba)."));
+  }
+
+#if NET5_0_OR_GREATER
+  [Test]
+  public void Contains_WithReadOnlySetOfT ()
+  {
+    var possibleItems = SetupCollection<IReadOnlySet<Guid>>([ (Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value ]);
+
+    var query = QueryFactory.CreateQuery(
+        new QueryDefinition("Test", TestDomainStorageProviderDefinition, "SELECT * FROM [Order] o WHERE o.[ID] IN (SELECT [Value] FROM @1)", QueryType.Collection),
+        new QueryParameterCollection { { "@1", possibleItems } });
+
+    var queryResult = TestableClientTransaction.QueryManager.GetCollection<Order>(query);
+    CheckQueryResult(queryResult, DomainObjectIDs.Order1, DomainObjectIDs.Order3);
+  }
+
+  [Test]
+  public void Contains_WithReadOnlySetOfT_UsesTableTypeWithUniqueConstraint ()
+  {
+    // set up a set with duplicates to get a database exception when we insert these duplicates into a table with unique constraint
+    var possibleItems = SetupCollection<IReadOnlySet<Guid>>([(Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value]);
+
+    var query = QueryFactory.CreateQuery(
+        new QueryDefinition("Test", TestDomainStorageProviderDefinition, "SELECT * FROM [Order] o WHERE o.[ID] IN (SELECT [Value] FROM @1)", QueryType.Collection),
+        new QueryParameterCollection { { "@1", possibleItems } });
+
+    Assert.That(() => TestableClientTransaction.QueryManager.GetCollection<Order>(query).ToArray(),
+        Throws.InstanceOf<RdbmsProviderException>()
+            .With.Message.Contains("Error while executing SQL command: Violation of UNIQUE KEY constraint")
+            .With.Message.Contains("Cannot insert duplicate key in object 'dbo.@1'. The duplicate key value is (83445473-844a-4d3f-a8c3-c27f8d98e8ba)."));
+  }
+#endif
+
+  [Test]
+  public void Contains_UsingTableValuedParameter_NullValue ()
+  {
+    var orderNumbers = new int?[] { 2, null, 99 };
```

Comment:
Reminder needed: we discussed the option of having null values and updated the type in the actual tvp type in the database. Why did we decide against supporting it as a value in memory?

Reply:
What would the null value be used for? Keep in mind that this is about all columns in the row being `NULL`, because the collection item being `null` is what causes the exception.

At the moment, we only support TVP for `WHERE IN` clauses, where `NULL` just does not work:
```
declare @tbl table ([Value] int NULL)
insert into @tbl values (NULL)
select * from [TableWithAllDataTypes] where NaInt32WithNullValue IN (SELECT [Value] FROM @tbl)
```
does not return any row, even though there are 2 rows where `NaInt32WithNullValue` is `NULL`

For `JOIN` clauses, `NULL`-rows hardly make sense because there is nothing to join on; even though we can write a join condition that joins on `NULL`, this seems like a purely hypothetical scenario.

For `FROM` clauses, `NULL`-rows may make some sense in some fringe scenarios, but any scenario where we select from a table-valued parameter without joining (and thus filtering out the `NULL`-rows, see above) seems completely redundant.

Reply:
As discussed, we do not want to support `null` items, but we should throw a `NotSupportedException` that explains to the developer why `null` items are not supported.